scripts/endoom-tool

```diff
@@ -0,0 +1,565 @@
+
+    command_group_me.add_argument(
+        "-j",
+        "--join",
+        metavar="JOIN-DIRECTORY",
+        help="Join command. Join the directory previously created by --split " +
+            "to form an ENDOOM lump.")
+
+    command_group_me.add_argument(
+        "-s",
+        "--split",
+        metavar="SPLIT-DIRECTORY",
+        help="Split command. Split the ENDOOM lump into foreground, " +
+            "background and text in the specified directory.")
+
+    # Options that modify the command. Sorted alphabetically.
+
+    option_group = parser.add_argument_group("Options", "Options that modify " +
+                   "command behavior")
+
+    option_group.add_argument(
+        "-c",
+        "--clean",
+        action="store_true",
+        help="Clean. Make foreground equal to background for spaces, and convert " +
+              "to space when foreground and background is the same color. " +
+              "Recommended when the exact ENDOOM does not need to be maintained. " +
+              "This option has no effect with -j, --join.")
+
+    option_group.add_argument(
+        "-p",
+        "--plain",
+        action="store_true",
+        help="Plain. Disable all ANSI color effects. For -j, --join this means " +
+             "to use white and black instead of what's in the foreground and " +
+             "background files before any other color processing.")
+
+    option_group.add_argument(
+        "-q",
+        "--quiet",
+        action="store_true",
+        help="Quiet. Disable some warnings and noise.")
+
+    option_group.add_argument(
+        "-r",
+        "--random-colors",
+        action="store_true",
+        help="Random colors. Make the colors a hash of two bytes associated " +
+             "with each character in order to make it easier to see " +
+             "inconsistencies that are otherwise hidden.")
+
+    option_group.add_argument(
+        "-t",
+        "--tolerant",
+        action="store_true",
+        help="Tolerate missing data. Missing files and data are considered " +
+              "to be black spaces.")
+
+    parser.add_argument(
+        "endooms",
+        metavar="ENDOOM",
+        nargs="+",
+        help="ENDOOM lump to process.")
+
+    args = parser.parse_args()
+
+    return args
+
+# Process an ENDOOM lump.
+def process_endoom(endoom_path):
+    # If true then there were one or more nulls in the input.
+    null_mapped_to_space = False
+    nbsp_mapped_to_space = False
+    black_space = b" \x00"
+    try:
+        read_write_binary = "wb" if args.join else "rb"
+        endoom_hand = open_file(endoom_path, read_write_binary)
+    except OSError:
+        if args.tolerant:
+            endoom_hand = None
+        else:
+            print("ENDOOM", endoom_path, "could not be opened for",
+                ("write" if args.join else "read", ), ".", file=sys.stderr)
+            sys.exit(1)
+    if args.join or args.split:
```

Comment:
You have a lot of these `if args.foo` lines throughout the file (24, if my search box is to be believed). It's a good indication that the code is factored wrong.

This function (`process_endoom`) doesn't have a clear purpose as it currently stands. The name is a good indicator of this - it's basically "do what this script does". As a suggestion, try making four copies of this function named `join()`, `split()`, `clean()` and `display()`, and then go through each and remove all the code that doesn't apply to that particular one. For `clean()` and `display()` you'd be able to delete this block, for example, and then for `join()` and `split()` you don't need the `if`.

You might find that the result is much clearer. If there's duplicated code you can then factor that out into common functions that they use.

Reply:
One possibility is that we commit this as is now and you can make those changes as you see fit. Or we can close this PR and I can just add `endoom-tool` to one of my personal repos.